File: pr-preview/pr-75/community/training/debugging/perfect_numbers.py
#!/usr/bin/env python3
"""
This script prints perfect numbers.

Perfect numbers are positive integers that are equal to the sum of their
divisors.
"""


def main():
    start = 2
    end = 1_000
    for value in range(start, end + 1):
        if is_perfect(value):
            print(value)


def is_perfect(value):
    divisors = divisors_of(value)
    return sum(divisors) == value


def divisors_of(value):
    divisors = []
    candidate = 2
    while candidate < value:
        if value % candidate == 0:
            divisors.append(candidate)
        candidate += 1
    return divisors


if __name__ == '__main__':
    main()

File: community/training/debugging/perfect_numbers.R
#!/usr/bin/env -S Rscript --vanilla
#
# This script prints perfect numbers.
#
# Perfect numbers are positive integers that are equal to the sum of their
# divisors.
#


main <- function() {
  start <- 2
  end <- 1000
  for (value in seq.int(start, end)) {
    if (is_perfect(value)) {
      print(value)
    }
  }
}


is_perfect <- function(value) {
  divisors <- divisors_of(value)
  sum(divisors) == value
}


divisors_of <- function(value) {
  divisors <- c()
  candidate <- 2
  while (candidate < value) {
    if (value %% candidate == 0) {
      divisors <- c(divisors, candidate)
    }
    candidate <- candidate + 1
  }
  divisors
}


main()
